File: levanta/ios/Levanta/Camera/AnalisadorDePose.swift
import AVFoundation
import Vision

/// Traduz cada quadro da câmera num `Esqueleto` e entrega ao chamador.
///
/// O Vision devolve juntas em coordenadas normalizadas com a origem no canto
/// **inferior** esquerdo; o motor de contagem trabalha com a origem no canto
/// superior, como o ML Kit. A inversão de `y` acontece aqui, e é toda a
/// diferença entre as duas plataformas.
final class AnalisadorDePose: NSObject, AVCaptureVideoDataOutputSampleBufferDelegate {

    private let aoDetectar: (Esqueleto) -> Void
    private let pedido = VNDetectHumanBodyPoseRequest()
    private let fila = DispatchQueue(label: "levanta.pose")

    init(aoDetectar: @escaping (Esqueleto) -> Void) {
        self.aoDetectar = aoDetectar
    }

    func captureOutput(_ output: AVCaptureOutput,
                       didOutput amostra: CMSampleBuffer,
                       from conexao: AVCaptureConnection) {
        guard let buffer = CMSampleBufferGetImageBuffer(amostra) else { return }
        let largura = Float(CVPixelBufferGetWidth(buffer))
        let altura = Float(CVPixelBufferGetHeight(buffer))

        let manipulador = VNImageRequestHandler(cvPixelBuffer: buffer, orientation: .up, options: [:])
        do {
            try manipulador.perform([pedido])
            guard let observacao = pedido.results?.first else { return }
            if let esqueleto = converter(observacao, largura: largura, altura: altura) {
                aoDetectar(esqueleto)
            }
        } catch {
            // Um quadro perdido não é erro: o próximo chega em 33 ms.
        }
    }

    private func converter(_ o: VNHumanBodyPoseObservation,
                           largura: Float, altura: Float) -> Esqueleto? {
        guard let juntas = try? o.recognizedPoints(.all) else { return nil }
        var marcos: [Marco: Ponto] = [:]
        for (meu, doVision) in Self.equivalencias {
            guard let j = juntas[doVision], j.confidence > 0 else { continue }
            marcos[meu] = Ponto(
                x: Float(j.location.x),
                // O Vision cresce `y` para cima; o motor espera para baixo.
                y: 1 - Float(j.location.y),
                confianca: Float(j.confidence),
            )
        }
        let agora = Int64(Date().timeIntervalSince1970 * 1000)
        return Esqueleto(marcos: marcos, proporcao: largura / altura, instanteMs: agora)
    }

    private static let equivalencias: [Marco: VNHumanBodyPoseObservation.JointName] = [
        .nariz: .nose,
        .ombroEsq: .leftShoulder, .ombroDir: .rightShoulder,
        .cotoveloEsq: .leftElbow, .cotoveloDir: .rightElbow,
        .pulsoEsq: .leftWrist, .pulsoDir: .rightWrist,
        .quadrilEsq: .leftHip, .quadrilDir: .rightHip,
        .joelhoEsq: .leftKnee, .joelhoDir: .rightKnee,
        .tornozeloEsq: .leftAnkle, .tornozeloDir: .rightAnkle,
    ]
}

File: levanta/android/app/src/main/kotlin/br/com/levanta/camera/AnalisadorDePose.kt
package br.com.levanta.camera

import androidx.annotation.OptIn
import androidx.camera.core.ExperimentalGetImage
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import br.com.levanta.core.pose.Esqueleto
import br.com.levanta.core.pose.Marco
import br.com.levanta.core.pose.Ponto
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.pose.Pose
import com.google.mlkit.vision.pose.PoseDetection
import com.google.mlkit.vision.pose.PoseLandmark
import com.google.mlkit.vision.pose.defaults.PoseDetectorOptions

/**
 * Traduz cada quadro da câmera num [Esqueleto] e entrega ao chamador.
 *
 * O ML Kit devolve 33 marcos em pixels da imagem original; aqui eles viram os
 * 13 marcos normalizados que o motor de contagem entende, no mesmo formato que
 * o adaptador do iOS produz a partir do Vision. Toda a diferença entre as duas
 * plataformas termina neste arquivo.
 */
class AnalisadorDePose(
    private val aoDetectar: (Esqueleto) -> Unit,
) : ImageAnalysis.Analyzer {

    private val detector = PoseDetection.getClient(
        PoseDetectorOptions.Builder()
            // STREAM_MODE reaproveita o resultado anterior para rastrear entre
            // quadros: é bem mais estável em vídeo que SINGLE_IMAGE_MODE, e a
            // estabilidade é o que impede o contador de tremer no limiar.
            .setDetectorMode(PoseDetectorOptions.STREAM_MODE)
            .build(),
    )

    @OptIn(ExperimentalGetImage::class)
    override fun analyze(imagem: ImageProxy) {
        val quadro = imagem.image
        if (quadro == null) { imagem.close(); return }

        val giro = imagem.imageInfo.rotationDegrees
        val entrada = InputImage.fromMediaImage(quadro, giro)
        // Depois do giro, largura e altura trocam de lugar em 90° e 270°.
        val largura = if (giro == 90 || giro == 270) imagem.height else imagem.width
        val altura = if (giro == 90 || giro == 270) imagem.width else imagem.height

        detector.process(entrada)
            .addOnSuccessListener { pose ->
                aoDetectar(converter(pose, largura.toFloat(), altura.toFloat()))
            }
            .addOnCompleteListener { imagem.close() }
    }

    fun encerrar() = detector.close()

    private fun converter(pose: Pose, largura: Float, altura: Float): Esqueleto {
        val marcos = HashMap<Marco, Ponto>(EQUIVALENCIAS.size)
        for ((meu, doMlKit) in EQUIVALENCIAS) {
            val m = pose.getPoseLandmark(doMlKit) ?: continue
            marcos[meu] = Ponto(
                x = m.position.x / largura,
                y = m.position.y / altura,
                confianca = m.inFrameLikelihood,
            )
        }
        return Esqueleto.de(marcos, largura / altura, System.currentTimeMillis())
    }

    private companion object {
        val EQUIVALENCIAS = mapOf(
            Marco.NARIZ to PoseLandmark.NOSE,
            Marco.OMBRO_ESQ to PoseLandmark.LEFT_SHOULDER,
            Marco.OMBRO_DIR to PoseLandmark.RIGHT_SHOULDER,
            Marco.COTOVELO_ESQ to PoseLandmark.LEFT_ELBOW,
            Marco.COTOVELO_DIR to PoseLandmark.RIGHT_ELBOW,
            Marco.PULSO_ESQ to PoseLandmark.LEFT_WRIST,
            Marco.PULSO_DIR to PoseLandmark.RIGHT_WRIST,
            Marco.QUADRIL_ESQ to PoseLandmark.LEFT_HIP,
            Marco.QUADRIL_DIR to PoseLandmark.RIGHT_HIP,
            Marco.JOELHO_ESQ to PoseLandmark.LEFT_KNEE,
            Marco.JOELHO_DIR to PoseLandmark.RIGHT_KNEE,
            Marco.TORNOZELO_ESQ to PoseLandmark.LEFT_ANKLE,
            Marco.TORNOZELO_DIR to PoseLandmark.RIGHT_ANKLE,
        )
    }
}
